File: client/src/main/java/org/dreamexposure/discal/client/module/misc/StatusChanger.java
package org.dreamexposure.discal.client.module.misc;

import discord4j.core.object.presence.Activity;
import discord4j.core.object.presence.Presence;
import org.dreamexposure.discal.client.DisCalClient;
import org.dreamexposure.discal.core.database.DatabaseManager;
import org.dreamexposure.discal.core.object.BotSettings;
import org.dreamexposure.discal.core.utils.GlobalConst;

import java.util.ArrayList;
import java.util.TimerTask;

/**
 * Created by Nova Fox on 3/5/2017.
 * Website: www.cloudcraftgaming.com
 * For Project: DisCal
 */
public class StatusChanger extends TimerTask {
    private final ArrayList<String> statuses = new ArrayList<>();
    private Integer index;

    /**
     * Creates the StatusChanger and its Statuses list.
     */
    public StatusChanger() {
        this.statuses.add("Discord Calendar");
        this.statuses.add("!help for help");
        this.statuses.add("!DisCal for info");
        this.statuses.add("Powered by DreamExposure");
        this.statuses.add("Ultimate calendar bot!");
        this.statuses.add("%guCount% guilds on shard!");
        this.statuses.add("%calCount% calendars!");
        this.statuses.add("%annCount% announcements!");
        this.statuses.add("%shards% shards!");
        this.statuses.add("Version " + GlobalConst.version);
        this.statuses.add("DisCal is on Patreon!");
        this.index = 0;
    }

    @Override
    public void run() {
        if (DisCalClient.getClient() != null) {

            String status = this.statuses.get(this.index);
            status = status.replace("%guCount%", DisCalClient.getClient().getGuilds().count().block() + "");
            status = status.replace("%calCount%", DatabaseManager.INSTANCE.getCalendarCount().block() + "");
            status = status.replace("%annCount%", DatabaseManager.INSTANCE.getAnnouncementCount().block() + "");
            status = status.replace("%shards%", BotSettings.SHARD_COUNT.get());

            DisCalClient.getClient().updatePresence(Presence.online(Activity.playing(status))).subscribe();
            //Set new index.
            if (this.index + 1 >= this.statuses.size())
                this.index = 0;
            else
                this.index++;
        }
    }
}

File: client/src/main/kotlin/org/dreamexposure/discal/client/service/StatusChanger.kt
package org.dreamexposure.discal.client.service

import discord4j.core.GatewayDiscordClient
import discord4j.core.`object`.presence.ClientActivity
import discord4j.core.`object`.presence.ClientPresence
import org.dreamexposure.discal.Application
import org.dreamexposure.discal.GitProperty
import org.dreamexposure.discal.core.database.DatabaseManager
import org.springframework.boot.ApplicationArguments
import org.springframework.boot.ApplicationRunner
import org.springframework.stereotype.Component
import reactor.core.publisher.Flux
import reactor.core.publisher.Mono
import reactor.function.TupleUtils
import java.time.Duration
import java.util.concurrent.atomic.AtomicInteger

@Component
class StatusChanger(
    private val discordClient: GatewayDiscordClient,
): ApplicationRunner {
    private val index = AtomicInteger(0)

    private val statuses = listOf(
            "Discord Calendar",
            "!help for help",
            "!DisCal for info",
            "Powered by DreamExposure",
            "{guilds} guilds on shard!",
            "{calendars} calendars managed!",
            "{announcements} announcements running!",
            "{shards} total shards!",
            "Version {version}",
            "DisCal is on Patreon!",
    )

    private fun update(): Mono<Void> {
        val guCountMono = discordClient.guilds.count()
        val calCountMono = DatabaseManager.getCalendarCount()
        val annCountMono = DatabaseManager.getAnnouncementCount()

        return Mono.zip(guCountMono, calCountMono, annCountMono)
                .flatMap(TupleUtils.function { guilds, calendars, announcements ->
                    val currentIndex = index.get()
                    //Update index
                    if (currentIndex + 1 >= statuses.size)
                        index.lazySet(0)
                    else
                        index.lazySet(currentIndex + 1)

                    //Get status we want to change to
                    val status = statuses[currentIndex]
                            .replace("{guilds}", guilds.toString())
                            .replace("{calendars}", calendars.toString())
                            .replace("{announcements}", announcements.toString())
                            .replace("{shards}", Application.getShardCount().toString())
                            .replace("{version}", GitProperty.DISCAL_VERSION.value)


                    discordClient.updatePresence(ClientPresence.online(ClientActivity.playing(status)))
                })
    }

    override fun run(args: ApplicationArguments?) {
        Flux.interval(Duration.ofMinutes(5))
                .flatMap { update() }
                .subscribe()
    }
}
